bpf: Support proxy upstream connections with original source address/port

jrajahalme opened the pull request:
Allow Envoy egress proxies to use the original source address and port in the upstream connections. This allows destination node to map the source IP to the policy ID. Egress Kafka is not supported yet, as Kafka only works for ingress at the moment. DNS proxy (always egress) support for original source address is left for a follow-up PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8870)
<!-- Reviewable:end -->

joestringer commented on pkg/datapath/iptables/iptables.go:
```diff
@@ -552,7 +552,41 @@ func (m *IptablesManager) InstallRules(ifName string) error {
+			m.waitArgs,
+			"-t", "mangle",
+			"-A", ciliumPreMangleChain,
+			"-m", "socket", "--transparent", "--nowildcard",
```
(Note, not a question): These rules overlap with the same ones being added in https://github.com/cilium/cilium/pull/8864.

jrajahalme replied:
Right, will rebase on top of #8864 when it merges.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -460,6 +466,12 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 	reason = ret;
 
+	// Check it this is return traffic to an ingress proxy.
```
The comment says "ingress proxy" but the code below is setting the mark for egress proxy too..?[0]

jrajahalme replied:
reply direction traffic from an LXC, if going to a local proxy, is for the ingress proxy. Original direction traffic from an LXC, if going to a local proxy, if for the egress proxy :-)

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -500,7 +512,7 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 		ct_state_new.src_sec_id = SECLABEL;
 		ret = ct_create4(get_ct_map4(&tuple), &tuple, skb, CT_EGRESS,
-				 &ct_state_new);
+				 &ct_state_new, verdict > 0);
```
[0] This is in handling packets "from_lxc", ie from the endpoint, so the verdict here (if nonzero) is an egress L7 policy.

jrajahalme replied:
See above

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -790,6 +802,14 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, __u8 *reason, s
+	// Check it this is return traffic to an egress proxy.
+	// Do not redirect again if the packet is coming from the egress proxy.
+	if ((ret == CT_REPLY || ret == CT_RELATED) && ct_state.proxy_redirect &&
+	    !tc_index_skip_egress_proxy(skb)) {
```
I'm confused. This function is `ipv6_policy()` which is called from `tail_ipv6_policy()` and `tail_ipv6_to_endpoint()`, ie the endpoint ingress case. Why is an egress proxy skip bit being checked here in the ingress path?

jrajahalme replied:
See above.

joestringer reviewed (changes requested):
I took a quick look over the changes, and I'm a bit surprised by some of them, questions below.

jrajahalme committed:
bpf: Support proxy using original source address and port.

Add a 'proxy_redirect' bit to the conntrack entry so that the reply
direction packets on proxy upstream connections using the original
source address and port in addition to the original destination
address and port can be redirected back to the local stack for local
delivery.

iptables rules are added to mark packets matching a transparent socket
as going to the host proxy.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
